scipy/signal/ltisys.py

```diff
@@ -2494,3 +2497,271 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+#        OK even if it is useless
+        P_update_full = np.zeros(
+            (H_start + P_update.shape[0], H_start + P_update.shape[1])
+        )
+        P_update_full[:H_start, :H_start] = np.eye(H_start)
+        P_update_full[H_start:, H_start:] = P_update
+
+        if P is None:  # P1
+            P = P_update_full
+        else:
+            P = np.dot(P_update_full, P)  # "update P as above page 175"
+
+        # update H and extract the new "H_k_k-1" we will work on
+        H = np.dot(np.dot(P, A), P.T)
+        H_sub = H[control_index:, H_start:control_index]
+        if np.allclose(H_sub, 0):  # or H_k_k-1" is a zero matrix
+            stop = True
+
+    Bbar = np.dot(P, B)
+    if C is None:
+        Cbar = None
+    else:
+        Cbar = np.dot(C, P.T)
+
+    hessenberg_form = Bunch()
+    hessenberg_form.Abar = H
+    hessenberg_form.Bbar = Bbar
+    hessenberg_form.Cbar = Cbar
+    hessenberg_form.P = P
+    hessenberg_form.control_index = control_index
+    return hessenberg_form
+
+
+def controllability_matrix(A, B):
+    r"""
+    Compute the controllability matrix of the linear system :math:`AX+BU`.
+
+    Compute the controllability matrix of the linear system :math:`AX+BU`.
+    If a linear system is controllable it is possible to find a set of inputs
+    U such as any configuration of the state-vector X is reachable in finite
+    time starting from any other configuration.
+    If the rank of the controllability matrix is not equal to the dimension of
+    the state-space the system is not controllable.
+
+    Parameters
+    ----------
+    A, B : ndarray
+        State-space representation of linear system :math:`AX+BU`.
+
+    Returns
+    -------
+    Ctrb_mat : ndarray
+        The controllability matrix
+
+    Notes
+    -----
+    Even if the rank of the controllability matrix equals the dimension of the
+    state-space the system might be very difficult, if not impossible, to
+    control, its condition number should also be checked. **One should only
+    rely on the rank of this matrix for educational purposes due to the
+    numerical instability of this computation.**
+    For serious control design `controller_hessenberg` is the way to go.
+    """
+    if A.ndim != 2:
+        raise ValueError("A must be a 2D array/matrix.")
+    if B.ndim != 2:
+        raise ValueError("B must be a 2D array/matrix")
+    if A.shape[0] != A.shape[1]:
+        raise ValueError("A must be square")
+
+    Ctrb_mat = np.zeros((A.shape[0], A.shape[0] * B.shape[1]))
+    Ctrb_mat[:, 0:B.shape[1]] = B
+    for dim in range(1, A.shape[0]):
+        Ctrb_mat[:, dim * B.shape[1]:(dim + 1) * B.shape[1]] = np.dot(
+            A, Ctrb_mat[:, (dim - 1) * B.shape[1]:dim * B.shape[1]])
+    return Ctrb_mat
+
+
+def observability_matrix(A, C):
+    r"""
+    Compute the observability matrix of a linear system.
+
+    Compute the observability matrix of the linear system:
```

Comment:
Instead of repeating the same text here, you can say `The observability is defined as:`